### DISCUSSION TITLE
`Maximum update depth exceeded` error when creating a store with effect to store a `react-query` query

### DISCUSSION THREAD

**Discussion by donrsh** [pmndrs/zustand#1021] [Q&A] [answered]
I tried to wrap a store with effect that store a `useQuery` returned value, therefore I can use the  instance to refetch the query anywhere.

I found something interesting, though. Here's what I've done:
1. Define a zustand store (`useDataStore`) to store the query and its data.
2. Define a hook (`usePreparedDataStore`). Within it, I declare a query instance that auto fetch, and an effect to sync the query instance. The hook reexport the zustand store for convenience.

(Here's the [codesandbox link](https://codesandbox.io/s/zustand-and-react-query-maximum-update-v8jlk1?file=/src/App.tsx))

Now, two use cases:
Case 1. Invoke `usePreparedDataStore` in a parent component, and consume the store in a child component.
Case 2. Invoke `usePreparedDataStore` and consume store in the returned directly, in a component.

Case 1 works fine. However, the `Maximum update depth exceeded` error occurs in case 2.
(You can toggle the radio button to see the difference. )

Is it a bug? Or I misunderstand something about zustand? Since they use the same hook, I wonder it has things to do with react-query. In fact, one can comment the zustand set function invoke in the effect, and it turns out that there's no error in case 2.

**Comment by dai-shi** [comment, marked as answer]
It looks like that `dataQuery` is a new object in each render, and you pass that object to useEffect deps. I think it's a general React usage issue.

**Reply by donrsh** [reply]
Yes, I observe that, too. The problem is: why would that happen?

As I mentioned, here a try: comment the `set` call within the effect (line 37-40 in `useData.ts`), the error is gone. See the log, I log some message when the effect is triggered, you can see it triggered only 3 times (as expected).

Updated: As long as `dataQuery` is not set (i.e. set `data` only), there is no error.

**Reply by dai-shi** [reply]
Because your selector is this:
```js
    const props = useDataStore(
      (x) => ({ dataQuery: x.dataQuery, data: x.data }),
      shallow
    );
```
If `dataQuery` is undefined, and `data` is undefined, the shallow comparison tells unchanged, thus no infinite loop.

**Reply by donrsh** [reply]
I assume even if I set `dataQuery` in `useEffect`, there SHOULD NOT be infinite loop. The `dataQuery` should reach stable after 3 updates, as Case 1 (separate prepare) shows (open the console, there are 3 logs).

<img width="402" alt="image" src="https://user-images.githubusercontent.com/7490144/174450133-bf08c140-012b-44ad-bc85-b20d3f81388f.png">

**Reply by dai-shi** [reply]
Here's a reproduction without zustand.

https://codesandbox.io/s/boring-pine-ubutgf?file=/src/App.js

**Reply by donrsh** [reply]
Wow! OK I got it. Thanks for your answer. I would separate the fetch and consume logic in practice.
